katdal/caldata.py

```diff
@@ -0,0 +1,383 @@
+        # Interpolate the calibration solutions for the selected range
+        for key in katdal_obj._cal_solns.keys():
+            if katdal_obj._cal_solns[key]['interp'] is not None:
+                # interpolate over values, or repeat single value
+                # expected size <ts><ch><pol><ant>
+                if hasattr(katdal_obj._cal_solns[key]['interp'], '__call__'):
+                    solns = katdal_obj._cal_solns[key]['interp'](timestamps)
+                else:
+                    solns = np.repeat(katdal_obj._cal_solns[key]['interp'],
+                                      timestamps.size,
+                                      axis=0)
+                if key == 'K':
+                    solns = np.exp(solns * katdal_obj._delay_to_phase)
+                katdal_obj._cal_solns[key]['solns'] = da.from_array(solns,
+                                                                    chunks=(ts_chunk_size, ch_chunk_size, 1, 1),
+                                                                    )
+
+    def _cal_calc(katdal_obj):
+        """Calculate calibration and weight coefficients"""
+        katdal_obj._cal_coeffs = []
+        katdal_obj._wght_coeffs = []
+
+        _cal_setup(katdal_obj)
+        _cal_interp(katdal_obj.timestamps)
+
+        _bls_len = katdal_obj._corrprod_keep.sum()
+        _chan_len = katdal_obj._freq_keep.sum()
+        _time_len = katdal_obj._time_keep.sum()
+
+        if _bls_len != len(katdal_obj._cp_lookup):
+            raise ValueError('Shape mismatch between correlation products.')
+        if _chan_len != len(katdal_obj._data_channel_freqs):
+            raise ValueError('Shape mismatch in frequency axis.')
+        if _time_len != len(katdal_obj.timestamps):
+            raise ValueError('Shape mismatch in timestamps.')
+
+        # calibrate visibilities
+        K = katdal_obj._cal_solns['K']['solns']
+        B = katdal_obj._cal_solns['B']['solns']
+        G = katdal_obj._cal_solns['G']['solns']
+
+        # ((X*K)/B)/G
+        for idx, cp in enumerate(katdal_obj._cp_lookup):
+            scale_coeff = None
+            scale_wght = None
+            # K
+            scale_coeff = K[:, :, cp[0][0], cp[0][1]] * K[:, :, cp[1][0], cp[1][1]].conj()
+            # B
+            scale = B[:, :, cp[0][0], cp[0][1]] * B[:, :, cp[1][0], cp[1][1]].conj()
+            scale_coeff /= scale
+            scale_wght = (scale.real**2 + scale.imag**2)
+            # G
+            scale = G[:, :, cp[0][0], cp[0][1]] * G[:, :, cp[1][0], cp[1][1]].conj()
+            scale_coeff *= np.reciprocal(scale)
+            scale_wght *= (scale.real**2 + scale.imag**2)
+            katdal_obj._cal_coeffs.append(scale_coeff)
+            katdal_obj._wght_coeffs.append(scale_wght)
+
+        # coefficient matrices <ts><ch><bl>
+        katdal_obj._cal_coeffs = da.stack(katdal_obj._cal_coeffs, axis=2)
+        katdal_obj._wght_coeffs = da.stack(katdal_obj._wght_coeffs, axis=2)
+        return katdal_obj
```

Comment:
Indeed it if very inefficient, but the calculation is done per baseline and then needs to be reconstructed.
I have updated the calculation so that this part only gets done once. That creates big dask matrices, but at least from then the access is a little smoother.

I am unsure if this can be optimised more without affecting the calculation, and slightly unwilling to do so without input from TomM.